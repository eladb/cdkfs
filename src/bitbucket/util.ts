

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}