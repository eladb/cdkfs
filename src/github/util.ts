export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
