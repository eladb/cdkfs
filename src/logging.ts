import * as chalk from 'chalk';

let enabled = true;

function log(color: chalk.ChalkFunction, ...text: any[]) {
  if (!enabled) { return; }
  console.error(`ðŸ¤– ${color(...text)}`);
}

/**
 * @deprecated
 */
export function debug(...text: any[]) {
  if (process.env.DEBUG) {
    log(chalk.gray, ...text);
  }
}

/**
 * @deprecated
 */
export function verbose(...text: any[]) {
  log(chalk.white, ...text);
}

/**
 * @deprecated
 */
export function info(...text: any[]) {
  log(chalk.cyan, ...text);
}

/**
 * @deprecated
 */
export function error(...text: any[]) {
  log(chalk.red, ...text);
}

/**
 * @deprecated
 */
export function warn(...text: any[]) {
  log(chalk.yellow, ...text);
}

/**
 * @deprecated
 */
export function disable() {
  enabled = false;
}
