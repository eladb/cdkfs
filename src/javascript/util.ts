import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve } from "path";
import * as semver from "semver";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/.]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
