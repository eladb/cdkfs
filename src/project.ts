import * as chalk from 'chalk';
import { cleanup } from './cleanup';
import { printStartMenu } from './cli/cmds/start-app';
import { Component } from './component';
import { IgnoreFile } from './ignore-file';
import * as logging from './logging';
import { SampleReadme } from './readme';
import { Start } from './start';

/**
 * Base project
 */
export class Project {
  public readonly gitignore: IgnoreFile;

  private readonly components = new Array<Component>();
  private readonly tips = new Array<string>();

  constructor() {
    this.gitignore = new IgnoreFile(this, '.gitignore');

    new SampleReadme(this, '# my project');
  }

  /**
   * Prints a "tip" message during synthesis.
   * @param message The message
   */
  public addTip(message: string) {
    this.tips.push(message);
  }

  /**
   * Synthesize all project files into `outdir`
   * @param outdir The project root directory (default is `.`).
   */
  public synth(outdir: string = '.'): void {
    this.preSynthesize(outdir);

    // delete all generated files before we start synthesizing new ones
    cleanup(outdir);

    for (const comp of this.components) {
      comp.synthesize(outdir);
    }

    for (const comp of this.components) {
      comp.postSynthesize(outdir);
    }

    // project-level hook
    this.postSynthesize(outdir);


    logging.info('Synthesis complete');

    const start = this.components.find(c => c instanceof Start);
    if (start) {
      console.error();
      console.error('-'.repeat(100));
      printStartMenu(outdir);
    }

    if (this.tips.length) {
      console.error(chalk.cyanBright.underline('Tips:'));
      for (const tip of this.tips) {
        console.error(`ðŸ’¡ ${tip}`);
      }
    }
  }

  public preSynthesize(_outdir: string) {}

  /**
   * Called after synthesis. Order is *not* guaranteed.
   * @param _outdir The project directory
   */
  public postSynthesize(_outdir: string) {}

  /**
   * Adds a component to the project.
   * @internal
   */
  public _addComponent(component: Component) {
    this.components.push(component);
  }
}
