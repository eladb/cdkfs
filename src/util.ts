import * as child_process from 'child_process';
import * as path from 'path';
import * as fs from 'fs-extra';
import * as logging from './logging';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const decamelize = require('decamelize');

export function exec(command: string, options?: child_process.ExecSyncOptions) {
  logging.verbose(command);
  return child_process.execSync(command, {
    stdio: ['inherit', process.stderr, 'pipe'],
    ...options,
  });
}

/**
 * Executes `command` and returns its value or undefined if the command failed.
 */
export function execOrUndefined(command: string, options?: child_process.ExecSyncOptions): string | undefined {
  try {
    const value = child_process.execSync(command, { stdio: ['inherit', 'pipe', 'ignore'], ...options }).toString('utf-8').trim();
    if (!value) { return undefined; } // an empty string is the same as undefined
    return value;
  } catch {
    return undefined;
  }
}

export interface WriteFileOptions {
  /**
   * Whether the generated file should be marked as executable.
   *
   * @default false
   */
  executable?: boolean;

  /**
   * Whether the generated file should be readonly.
   *
   * @default false
   */
  readonly?: boolean;
}

export function getFilePermissions(options: WriteFileOptions): string {
  const readonly = options.readonly ?? false;
  const executable = options.executable ?? false;
  if (readonly && executable) {
    return '500';
  } else if (readonly) {
    return '400';
  } else if (executable) {
    return '755';
  } else {
    return '644';
  }
}

export function writeFile(filePath: string, data: any, options: WriteFileOptions = {}) {
  if (fs.existsSync(filePath)) {
    fs.chmodSync(filePath, '600');
  }

  fs.mkdirpSync(path.dirname(filePath));
  fs.writeFileSync(filePath, data);

  fs.chmodSync(filePath, getFilePermissions(options));
}

/**
 * Decamelizes the keys of an object structure, recursing through child objects and arrays.
 * @experimental
 */
export interface DecamelizeRecursivelyOptions {
  /**
   * Max depth to recurse before erroring.
   * @default 10
   */
  maxDepth?: number;

  /**
   * Returns true when a key should be decamelized
   * @default - all keys are decamelized
   */
  shouldDecamelize?: (path: string[], value: any) => boolean;

  /**
   * Separator for decamelizing.
   * @default "_"
   */
  separator?: string;

  /**
   * Current path.
   * @internal
   */
  path?: string[];
}

export function decamelizeKeysRecursively(input: any, opt?: DecamelizeRecursivelyOptions): any {
  const shouldAlwaysDecamelize = () => true;
  const shouldDecamelize = opt?.shouldDecamelize ?? shouldAlwaysDecamelize;
  const separator = opt?.separator ?? '_';
  const path_ = opt?.path ?? [];
  const maxDepth = opt?.maxDepth ?? 10;

  if (path_.length > maxDepth) {
    throw new Error('Decamelled too deeply - check that the input has no circular references');
  }

  if (Array.isArray(input)) {
    return input.map((k, i) => decamelizeKeysRecursively(k, {
      ...opt,
      path: [...path_, i.toString()],
    }));
  }

  if (typeof input === 'object' && input !== null) {
    const mappedObject: Record<string, any> = {};
    for (const [key, value] of Object.entries(input)) {
      const transformedKey = shouldDecamelize([...path_, key], value)
        ? decamelize(key, separator)
        : key;

      mappedObject[transformedKey] = decamelizeKeysRecursively(value, {
        ...opt,
        path: [...path_, key],
      });
    }

    return mappedObject;
  }

  return input;
}

/**
 * Returns false if value is unset or a falsey value, and true otherwise.
 * @param value an environment variable
 */
export function isTruthy(value: string | undefined): boolean {
  return !(value === undefined || ['null', 'undefined', '0', 'false', ''].includes(value.toLocaleLowerCase()));
}

/**
 * Type of a map mapping strings to some arbitrary type
 */
export type Obj<T> = { [key: string]: T };

/**
 * Return whether the given value is an object
 *
 * Even though arrays and instances of classes technically are objects, we
 * usually want to treat them differently, so we return false in those cases.
 */
export function isObject(x: any): x is Obj<any> {
  return x !== null && typeof x === 'object' && !Array.isArray(x)
    && x.constructor.name === 'Object';
}

/**
 * Recursively merge objects together
 *
 * The leftmost object is mutated and returned. Arrays are not merged
 * but overwritten just like scalars.
 *
 * If an object is merged into a non-object, the non-object is lost.
 *
 * `undefined`s will cause a value to be deleted if destructive is enabled.
 */
export function deepMerge(objects: Array<Obj<any> | undefined>, destructive: boolean = false) {
  function mergeOne(target: Obj<any>, source: Obj<any>) {
    for (const key of Object.keys(source)) {
      const value = source[key];

      if (isObject(value)) {
        // if the value at the target is not an object, override it with an
        // object so we can continue the recursion
        if (!isObject(target[key])) {
          target[key] = value;
        }
        mergeOne(target[key], value);

        // if the result of the merge is an empty object, it's because the
        // eventual value we assigned is `undefined`, and there are no
        // sibling concrete values alongside, so we can delete this tree.
        const output = target[key];
        if (isObject(output) && Object.keys(output).length === 0 && destructive) {
          delete target[key];
        }
      } else if (value === undefined && destructive) {
        delete target[key];
      } else if (typeof value !== 'undefined') {
        target[key] = value;
      }
    }
  }

  const others = objects.filter(x => x != null) as Array<Obj<any>>;

  if (others.length === 0) { return {}; }
  const into = others.splice(0, 1)[0];

  others.forEach(other => mergeOne(into, other));
  return into;
}

/*
 * Deduplicate values in a list, returning a new array.
 * @param array list of values
 */
export function dedupArray<T>(array: T[]): T[] {
  return array.filter((val, idx) => array.indexOf(val) === idx);
}

/**
 * Returns a sorted version of `x` or `undefined` if it is an empty array or object.
 */
export function sorted<T>(x: T) {
  if (x == null) { return undefined; }
  if (Array.isArray(x)) {
    if (x.length === 0) { return undefined; }
    return (x as unknown[]).sort();
  } else if (typeof x === 'object') {
    if (Object.keys(x).length === 0) { return undefined; }
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(x).sort(([l], [r]) => l.localeCompare(r))) {
      result[key] = value;
    }
    return result as T;
  } else {
    return x;
  }
}

export function formatAsPythonModule(name: string) {
  return name.replace(/-/g, '_').replace(/\./g, '_');
}
