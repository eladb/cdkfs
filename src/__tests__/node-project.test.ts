import { NodeProject, NodeProjectOptions, LogLevel } from '..';
import { DependencyType } from '../deps';
import * as logging from '../logging';
import { NodePackage, NpmAccess } from '../node-package';
import { Project } from '../project';
import { mkdtemp, synthSnapshot, TestProject } from './util';

logging.disable();

test('license file is added by default', () => {
  // WHEN
  const project = new TestNodeProject();

  // THEN
  expect(synthSnapshot(project).LICENSE).toContain('Apache License');
});

test('license file is not added if licensed is false', () => {
  // WHEN
  const project = new TestNodeProject({
    licensed: false,
  });

  // THEN
  const snapshot = synthSnapshot(project);
  expect(snapshot.LICENSE).toBeUndefined();
  expect(snapshot['.gitignore']).not.toContain('LICENSE');
  expect(snapshot['package.json'].license).toEqual('UNLICENSED');
});

describe('deps', () => {

  test('runtime deps', () => {
    // GIVEN
    const project = new TestNodeProject({
      deps: [
        'aaa@^1.2.3',
        'bbb@~4.5.6',
      ],
    });

    // WHEN
    project.addDeps('ccc');
    project.deps.addDependency('ddd', DependencyType.RUNTIME);

    // THEN
    const pkgjson = packageJson(project);
    expect(pkgjson.dependencies).toStrictEqual({
      aaa: '^1.2.3',
      bbb: '~4.5.6',
      ccc: '*',
      ddd: '*',
    });
    expect(pkgjson.peerDependencies).toStrictEqual({});
  });

  test('dev dependencies', () => {
    // GIVEN
    const project = new TestNodeProject({
      devDeps: [
        'aaa@^1.2.3',
        'bbb@~4.5.6',
      ],
    });

    // WHEN
    project.addDevDeps('ccc');
    project.deps.addDependency('ddd', DependencyType.TEST);
    project.deps.addDependency('eee@^1', DependencyType.DEVENV);
    project.deps.addDependency('fff@^2', DependencyType.BUILD);

    // THEN
    const pkgjson = packageJson(project);
    expect(pkgjson.devDependencies.aaa).toStrictEqual('^1.2.3');
    expect(pkgjson.devDependencies.bbb).toStrictEqual('~4.5.6');
    expect(pkgjson.devDependencies.ccc).toStrictEqual('*');
    expect(pkgjson.devDependencies.ddd).toStrictEqual('*');
    expect(pkgjson.devDependencies.eee).toStrictEqual('^1');
    expect(pkgjson.devDependencies.fff).toStrictEqual('^2');
    expect(pkgjson.peerDependencies).toStrictEqual({});
    expect(pkgjson.dependencieds).toBeUndefined();
  });

  test('peerDependencies', () => {
    // GIVEN
    const project = new TestNodeProject({
      peerDeps: [
        'aaa@^1.2.3',
        'bbb@~4.5.6',
      ],
    });

    // WHEN
    project.addPeerDeps('ccc');
    project.deps.addDependency('ddd', DependencyType.PEER);

    // THEN
    const pkgjson = packageJson(project);
    expect(pkgjson.peerDependencies).toStrictEqual({
      aaa: '^1.2.3',
      bbb: '~4.5.6',
      ccc: '*',
      ddd: '*',
    });

    // devDependencies are added with pinned versions
    expect(pkgjson.devDependencies.aaa).toStrictEqual('1.2.3');
    expect(pkgjson.devDependencies.bbb).toStrictEqual('4.5.6');
    expect(pkgjson.devDependencies.ccc).toStrictEqual('*');
    expect(pkgjson.devDependencies.ddd).toStrictEqual('*');
    expect(pkgjson.dependencieds).toBeUndefined();
  });

  test('peerDependencies without pinnedDevDep', () => {
    // GIVEN
    const project = new TestNodeProject({
      peerDependencyOptions: {
        pinnedDevDependency: false,
      },
      peerDeps: [
        'aaa@^1.2.3',
        'bbb@~4.5.6',
      ],
    });

    // WHEN
    project.addPeerDeps('ccc');
    project.deps.addDependency('ddd', DependencyType.PEER);

    // THEN
    const pkgjson = packageJson(project);
    expect(pkgjson.peerDependencies).toStrictEqual({
      aaa: '^1.2.3',
      bbb: '~4.5.6',
      ccc: '*',
      ddd: '*',
    });

    // sanitize
    ['jest', 'jest-junit', 'projen', 'standard-version'].forEach(d => delete pkgjson.devDependencies[d]);

    expect(pkgjson.devDependencies).toStrictEqual({});
    expect(pkgjson.dependencieds).toBeUndefined();
  });

  test('bundled deps are automatically added as normal deps', () => {
    // GIVEN
    const project = new TestNodeProject({
      bundledDeps: ['hey@2.1.1'],
    });

    // WHEN
    project.addBundledDeps('foo@^1.2.3');
    project.deps.addDependency('bar@~1.0.0', DependencyType.BUNDLED);

    // THEN
    const pkgjson = packageJson(project);
    expect(pkgjson.dependencies).toStrictEqual({
      hey: '2.1.1',
      foo: '^1.2.3',
      bar: '~1.0.0',
    });
    expect(pkgjson.bundledDependencies).toStrictEqual([
      'bar',
      'foo',
      'hey',
    ]);
  });
});

describe('npm publishing options', () => {
  test('defaults', () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const npm = new NodePackage(project, {
      packageName: 'my-package',
    });

    // THEN
    expect(npm.npmAccess).toStrictEqual(NpmAccess.PUBLIC);
    expect(npm.npmDistTag).toStrictEqual('latest');
    expect(npm.npmRegistry).toStrictEqual('registry.npmjs.org');
    expect(npm.npmRegistryUrl).toStrictEqual('https://registry.npmjs.org/');

    // since these are all defaults, publishConfig is not defined.
    expect(synthSnapshot(project)['package.json'].publishConfig).toBeUndefined();
  });

  test('scoped packages default to RESTRICTED access', () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const npm = new NodePackage(project, {
      packageName: 'scoped@my-package',
    });

    // THEN
    expect(npm.npmAccess).toStrictEqual(NpmAccess.RESTRICTED);

    // since these are all defaults, publishConfig is not defined.
    expect(packageJson(project).publishConfig).toBeUndefined();
  });

  test('non-scoped package cannot be RESTRICTED', () => {
    // GIVEN
    const project = new TestProject();

    // THEN
    expect(() => new NodePackage(project, {
      packageName: 'my-package',
      npmAccess: NpmAccess.RESTRICTED,
    })).toThrow(/"npmAccess" cannot be RESTRICTED for non-scoped npm package/);
  });

  test('custom settings', () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const npm = new NodePackage(project, {
      packageName: 'scoped@my-package',
      npmDistTag: 'next',
      npmRegistryUrl: 'https://foo.bar',
      npmAccess: NpmAccess.PUBLIC,
    });

    // THEN
    expect(npm.npmDistTag).toStrictEqual('next');
    expect(npm.npmRegistry).toStrictEqual('foo.bar');
    expect(npm.npmRegistryUrl).toStrictEqual('https://foo.bar/');
    expect(npm.npmAccess).toStrictEqual(NpmAccess.PUBLIC);
    expect(packageJson(project).publishConfig).toStrictEqual({
      access: 'public',
      registry: 'https://foo.bar/',
      tag: 'next',
    });
  });

  test('deprecated npmRegistry can be used instead of npmRegistryUrl and then https:// is assumed', () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const npm = new NodePackage(project, {
      packageName: 'scoped@my-package',
      npmRegistry: 'foo.bar.com',
    });

    // THEN
    expect(npm.npmRegistry).toStrictEqual('foo.bar.com');
    expect(npm.npmRegistryUrl).toStrictEqual('https://foo.bar.com/');
    expect(packageJson(project).publishConfig).toStrictEqual({
      registry: 'https://foo.bar.com/',
    });
  });
});

test('extend github release workflow', () => {
  const project = new TestNodeProject();

  project.releaseWorkflow?.addJobs({
    publish_docker_hub: {
      'runs-on': 'ubuntu-latest',
      'env': {
        CI: 'true',
      },
      'steps': [
        {
          name: 'Check out the repo',
          uses: 'actions/checkout@v2',
        },
        {
          name: 'Push to Docker Hub',
          uses: 'docker/build-push-action@v1',
          with: {
            username: '${{ secrets.DOCKER_USERNAME }}',
            password: '${{ secrets.DOCKER_PASSWORD }}',
            repository: 'projen/projen-docker',
            tag_with_ref: true,
          },
        },
      ],
    },
  });

  const workflow = synthSnapshot(project)['.github/workflows/release.yml'];
  expect(workflow).toContain('publish_docker_hub:\n    runs-on: ubuntu-latest\n');
  expect(workflow).toContain('username: ${{ secrets.DOCKER_USERNAME }}\n          password: ${{ secrets.DOCKER_PASSWORD }}');
});

function packageJson(project: Project) {
  return synthSnapshot(project)['package.json'];
}

class TestNodeProject extends NodeProject {
  constructor(options: Partial<NodeProjectOptions> = {}) {
    super({
      outdir: mkdtemp(),
      name: 'test-node-project',
      logging: {
        level: LogLevel.OFF,
      },
      defaultReleaseBranch: 'master',
      ...options,
    });
  }
}