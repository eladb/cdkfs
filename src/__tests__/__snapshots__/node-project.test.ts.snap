// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutableBuild will push changes to PR branches 1`] = `
Array [
  Object {
    "name": "Checkout",
    "uses": "actions/checkout@v2",
    "with": Object {
      "ref": "\${{ github.event.pull_request.head.ref }}",
      "repository": "\${{ github.event.pull_request.head.repo.full_name }}",
    },
  },
  Object {
    "name": "Install dependencies",
    "run": "yarn install --check-files --frozen-lockfile",
  },
  Object {
    "name": "Set git identity",
    "run": "git config user.name \\"Automation\\"
git config user.email \\"github-actions@github.com\\"",
  },
  Object {
    "name": "Build",
    "run": "npx projen build",
  },
  Object {
    "id": "git_diff",
    "name": "Check for changes",
    "run": "git diff --exit-code || echo \\"::set-output name=has_changes::true\\"",
  },
  Object {
    "if": "steps.git_diff.outputs.has_changes",
    "name": "Commit and push changes (if changed)",
    "run": "git add . && git commit -m \\"chore: self mutation\\" && git push origin HEAD:\${{ github.event.pull_request.head.ref }}",
  },
  Object {
    "env": Object {
      "GITHUB_TOKEN": "\${{ secrets.GITHUB_TOKEN }}",
    },
    "if": "steps.git_diff.outputs.has_changes",
    "name": "Update status check (if changed)",
    "run": "gh api -X POST /repos/\${{ github.event.pull_request.head.repo.full_name }}/check-runs -F name=\\"build\\" -F head_sha=\\"$(git rev-parse HEAD)\\" -F status=\\"completed\\" -F conclusion=\\"success\\"",
  },
]
`;
